homeassistant/components/qnap/__init__.py

```diff
@@ -1 +1,164 @@
 """The qnap component."""
+import asyncio
+from datetime import timedelta
+import logging
+
+import async_timeout
+from qnapstats import QNAPStats
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_MONITORED_CONDITIONS,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_SSL,
+    CONF_TIMEOUT,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+)
+from homeassistant.helpers import (
+    config_per_platform,
+    config_validation as cv,
+    device_registry as dr,
+)
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    COMPONENTS,
+    CONF_DRIVES,
+    CONF_NICS,
+    CONF_VOLUMES,
+    DEFAULT_PORT,
+    DEFAULT_TIMEOUT,
+    DOMAIN,
+)
+
+UPDATE_INTERVAL = timedelta(minutes=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass, config):
+    """Set up the qnap environment."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Import configuration from sensor platform
+    config_platform = config_per_platform(config, "sensor")
+    for p_type, p_config in config_platform:
+        if p_type != DOMAIN:
+            continue
+
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_IMPORT},
+                data=p_config,
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set the config entry up."""
+    if not config_entry.options:
+        options = {
+            CONF_VERIFY_SSL: config_entry.data.get(CONF_VERIFY_SSL, True),
+            CONF_SSL: config_entry.data.get(CONF_SSL, False),
+            CONF_TIMEOUT: config_entry.data.get(CONF_TIMEOUT, DEFAULT_TIMEOUT),
+            CONF_MONITORED_CONDITIONS: config_entry.data.get(CONF_MONITORED_CONDITIONS),
+            CONF_NICS: cv.ensure_list_csv(config_entry.data.get(CONF_NICS)),
+            CONF_DRIVES: cv.ensure_list_csv(config_entry.data.get(CONF_DRIVES)),
+            CONF_VOLUMES: cv.ensure_list_csv(config_entry.data.get(CONF_VOLUMES)),
+        }
+        hass.config_entries.async_update_entry(config_entry, options=options)
+
+    host = config_entry.data[CONF_HOST]
+    protocol = "https" if config_entry.options.get(CONF_SSL) else "http"
+    api = QNAPStats(
+        host=f"{protocol}://{host}",
+        port=config_entry.data.get(CONF_PORT, DEFAULT_PORT),
+        username=config_entry.data[CONF_USERNAME],
+        password=config_entry.data[CONF_PASSWORD],
+        verify_ssl=config_entry.options.get(CONF_VERIFY_SSL),
+        timeout=config_entry.options.get(CONF_TIMEOUT),
+    )
+    try:
+        system_info = await hass.async_add_executor_job(api.get_system_stats)
+    except:  # noqa: E722 pylint: disable=bare-except
```

Comment:
Please specify the errors that no internet connection and bad auth can cause. 

When it's no internet connection, raise `ConfigEntryNotReady`.